lib/shape.ml

```diff
@@ -4,74 +4,82 @@ type point = { x : int; y : int }
 type line = { a : point; b : point }
 type rectangle = { c : point; length : int; width : int }
 type circle = { c : point; radius : int }
-type ellipse = {c : point; rx : int; ry: int}
+type ellipse = { c : point; rx : int; ry : int }
 type shape = Circle of circle | Rectangle of rectangle | Ellipse of ellipse | Line of line
 type shapes = shape list
 
-let dimensions = ref {x = 500; y = 500}
-let set_dimensions x y = 
-  dimensions := {x;y} 
+let dimensions = ref { x = 500; y = 500 }
+let set_dimensions x y =
+  dimensions := { x; y }
 
-let canvas_mid = { x = (!dimensions.x / 2); y = (!dimensions.y / 2)}
+let canvas_center = { x = 0; y = 0 }
 
 let axes_flag = ref false
-let draw_axes flag = 
+let draw_axes flag =
   axes_flag := flag
 
-let draw_line x1 y1 x2 y2 = 
+let draw_line x1 y1 x2 y2 =
   draw_poly_line [|(x1, y1); (x2, y2)|]
 
+let denormalize x y =
+  { x = x + canvas_center.x; y = -y + canvas_center.y }
+
 let render_shape s =
   match s with
-  | Circle circle -> draw_circle circle.c.x circle.c.y circle.radius
+  | Circle circle ->
+      let c = denormalize circle.c.x circle.c.y in
+      draw_circle c.x c.y circle.radius
   | Rectangle rectangle ->
-      draw_rect rectangle.c.x rectangle.c.y rectangle.length rectangle.width
+      let c = denormalize rectangle.c.x rectangle.c.y in
+      draw_rect (c.x - (rectangle.length / 2)) (c.y - (rectangle.width / 2)) rectangle.length rectangle.width
   | Ellipse ellipse ->
-    draw_ellipse ellipse.c.x ellipse.c.y ellipse.rx ellipse.ry
-  | Line line -> draw_line line.a.x line.a.y line.b.x line.b.y
+      let c = denormalize ellipse.c.x ellipse.c.y in
+      draw_ellipse c.x c.y ellipse.rx ellipse.ry
+  | Line line ->
+      let a = denormalize line.a.x line.a.y in
+      let b = denormalize line.b.x line.b.y in
+      draw_line a.x a.y b.x b.y
 
 let circle ?x ?y r =
-  match (x, y) with
-  | Some x, Some y -> Circle { c = { x; y }; radius = r }
-  | _ -> Circle { c = { x = canvas_mid.x; y = canvas_mid.y }; radius = r }
+  let default_center = { x = 0; y = 0 } in
+  let center = match (x, y) with Some x, Some y -> { x; y } | _ -> default_center in
+  Circle { c = center; radius = r }
 
 let rectangle ?x ?y length width =
-  
   match (x, y) with
   | Some x, Some y -> Rectangle { c = { x; y }; length; width }
-  | _ -> Rectangle { c = { x = canvas_mid.x; y = canvas_mid.y }; length; width }
+  | _ -> Rectangle { c = { x = 0; y = 0 }; length; width }
 
 let ellipse ?x ?y rx ry =
-  match (x, y) with
-  | Some x, Some y -> Ellipse {c = {x; y}; rx; ry}
-  | _ -> Ellipse {c = { x = canvas_mid.x; y = canvas_mid.y}; rx; ry}
+  let default_center = { x = 0; y = 0 } in
+  let center = match (x, y) with Some x, Some y -> { x; y } | _ -> default_center in
+  Ellipse { c = center; rx; ry }
```

Comment:
I think the previous way to write this was easier to read. 
Please use that and instead of canvas_mid.x and canvas_mid.y, use 0.